pkg/authorization/admission/restrictusers/restrictusers.go

```diff
@@ -0,0 +1,158 @@
+var _ = oadmission.Validator(&restrictUsersAdmission{})
+
+// NewRestrictUsersAdmission configures an admission plugin that enforces
+// restrictions on adding role bindings in a project based on predefined
+// collaborators.
+func NewRestrictUsersAdmission(kclient kclientset.Interface) (admission.Interface, error) {
+	return &restrictUsersAdmission{
+		Handler: admission.NewHandler(admission.Create),
+		kclient: kclient,
+	}, nil
+}
+
+func (q *restrictUsersAdmission) SetOpenshiftClient(c oclient.Interface) {
+	q.oclient = c
+}
+
+// Admit makes admission decisions that enforce restrictions on adding
+// project-scoped role-bindings.  In order for a role binding to be permitted,
+// each subject in the binding must be matched by some collaborator matcher in
+// the namespace.
+func (q *restrictUsersAdmission) Admit(a admission.Attributes) (err error) {
+	if a.GetResource().GroupResource() != authorizationapi.Resource("rolebindings") {
+		return nil
+	}
+
+	// Ignore all operations that correspond to subresource actions.
+	if len(a.GetSubresource()) != 0 {
+		return nil
+	}
+
+	ns := a.GetNamespace()
+
+	// Ignore cluster-level resources.
+	if len(ns) == 0 {
+		return nil
+	}
+
+	binding, ok := a.GetObject().(*authorizationapi.RoleBinding)
+	if !ok {
+		return admission.NewForbidden(a,
+			fmt.Errorf("wrong object type for new rolebinding: %t", a.GetObject()))
+	}
+
+	oldbinding, ok := a.GetOldObject().(*authorizationapi.RoleBinding)
+	if oldbinding != nil && !ok {
+		return admission.NewForbidden(a,
+			fmt.Errorf("wrong object type for old rolebinding: %t", a.GetOldObject()))
+	}
+
+	glog.V(4).Infof("Handling rolebinding %s/%s", binding.Namespace, binding.Name)
+
+	// Only bother checking deltas.
+	newSubjects := []kapi.ObjectReference{}
+	if oldbinding == nil {
+		newSubjects = binding.Subjects
+	} else {
+		for _, newSubject := range binding.Subjects {
+			foundExisting := false
+			for _, existingSubject := range oldbinding.Subjects {
+				if existingSubject.Kind == newSubject.Kind &&
+					existingSubject.Name == newSubject.Name &&
+					existingSubject.Namespace == newSubject.Namespace {
+					foundExisting = true
+					break
+				}
+			}
+			if !foundExisting {
+				newSubjects = append(newSubjects, newSubject)
+			}
+		}
+	}
+
+	if len(newSubjects) == 0 {
+		glog.V(4).Infof("No new subjects; admitting")
+		return nil
+	}
+
+	// TODO: Cache collaborators.
+	matchersList, err := q.oclient.CollaboratorMatchers(ns).List(kapi.ListOptions{})
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+
+	if len(matchersList.Items) == 0 {
+		glog.V(4).Infof("No collaboration restrictions specified; admitting")
+		return nil
+	}
+
+	checker, err := NewRoleBindingRestrictions(ns, q.kclient, q.oclient)
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+
+	for _, matcher := range matchersList.Items {
+		checker.AddMatcher(&matcher)
+	}
+
+	// TODO: Aggregate errors and return an error that lists all subjects that are
```

Comment:
I see this TODO as required for merge.  Otherwise people will hate our API server even more.